RiotSwiftUI/Modules/Authentication/Registration/AuthenticationRegistrationViewModel.swift

```diff
@@ -0,0 +1,98 @@
+//
+// Copyright 2021 New Vector Ltd
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+import SwiftUI
+import Combine
+
+@available(iOS 14, *)
+typealias AuthenticationRegistrationViewModelType = StateStoreViewModel<AuthenticationRegistrationViewState,
+                                                                        Never,
+                                                                        AuthenticationRegistrationViewAction>
+
+
+@available(iOS 14, *)
+class AuthenticationRegistrationViewModel: AuthenticationRegistrationViewModelType, AuthenticationRegistrationViewModelProtocol {
+
+    // MARK: - Properties
+
+    // MARK: Public
+
+    @MainActor var completion: ((AuthenticationRegistrationViewModelResult) -> Void)?
+
+    // MARK: - Setup
+
+    init(homeserverAddress: String, showRegistrationForm: Bool = true, ssoIdentityProviders: [SSOIdentityProvider]) {
+        let bindings = AuthenticationRegistrationBindings()
+        let viewState = AuthenticationRegistrationViewState(homeserverAddress: HomeserverAddress.displayable(homeserverAddress),
+                                                            showRegistrationForm: showRegistrationForm,
+                                                            ssoIdentityProviders: ssoIdentityProviders,
+                                                            bindings: bindings)
+        
+        super.init(initialViewState: viewState)
+    }
+    
+    // MARK: - Public
+
+    override func process(viewAction: AuthenticationRegistrationViewAction) {
+        Task {
```

Comment:
Doesn't `Task` and `await MainActor.run` right after equal to just not using any concurrency? Or is the `process` method simetimes called from non-main thread? I wonder if that should be allowed if it is mean to accept view actions, i.e. actions triggered from the main thread.

Reply:
This is a side effect of marking `completion` as `@MainActor` where you get a compile error unless Swift can see that the call is also coming from the main thread. The right answer here is that  `process` should also be `@MainActor` but that breaks the protocol conformance, and so without updating the protocol and all of the SwiftUI view models this is the best "on-ramp" onto the `MainActor` I could come up with.

In reality the View Model (and Coordinator and View) should be annotated `@MainActor` and then none of the methods/properties would need annotating individually. Would be nice to get this right early on in EIX to avoid async calls like this.

Reply:
A couple of options that I could do to make this cleaner:
- Add a `@MainActor processOnMain(viewAction: AuthenticationRegistrationViewAction)` and call it from here.
- Add a `MainActorTask`/`TaskOnMain` that hides away the `await MainActor.run` the removes one level of closure.
Happy to consider either of these (or other ideas) if you think they make sense.

Reply:
`@MainActor private func` seems clean enough, though happy to go along the current code for this PR

Reply:
Perfect, will do that alongside renaming the `completions`.